For #3687 Use Metropolis Font for Header

kglazko opened the pull request:
---
<!-- Text above this line will be added to the commit once "bors" merges this PR -->

### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues finished by this pull request are added to the [milestone](https://github.com/mozilla-mobile/fenix/milestones) of the version currently in development.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

boek commented:
Hey @kglazko let's meet about this tomorrow and get this landed 👍

brampitoyo commented:
@kglazko @AmyYLee Just a note to say that @chrismsimpson has released a new version of Metropolis with “decent spacing & kerning” a few days ago.

If possible I recommend pulling the latest release before merging.

chrismsimpson commented:
You guys should definitely update, several things were fixed and new glyphs added, too.

kbrosnan commented:
Kate is on PTO through Monday

kglazko commented:
<img width="346" alt="Screen Shot 2020-02-19 at 5 58 27 PM" src="https://user-images.githubusercontent.com/1676235/74893816-198f7380-5342-11ea-9060-629e1016d3ce.png">
@brampitoyo @AmyYLee I changed the font for headers, please let me know if you would like me to specify a thinner or bolder look or other adjustments.

AmyYLee commented:
> <img alt="Screen Shot 2020-02-19 at 5 58 27 PM" width="346" src="https://user-images.githubusercontent.com/1676235/74893816-198f7380-5342-11ea-9060-629e1016d3ce.png">
> @brampitoyo @AmyYLee I changed the font for headers, please let me know if you would like me to specify a thinner or bolder look or other adjustments.

This looks good to me

AmyYLee commented:
@kglazko Reposting what was discussed on Slack:

• The buttons should be Metropolis - Bold - 14pt, 0.5 tracking
• CFR (ETP, PBM shortcut) component should be using Metropolis as body copy and buttons

kglazko commented:
<img width="385" alt="Screen Shot 2020-02-21 at 3 05 27 PM" src="https://user-images.githubusercontent.com/1676235/75078897-c2151300-54bb-11ea-9b1a-b151dff4d8ad.png">
<img width="386" alt="Screen Shot 2020-02-21 at 3 05 40 PM" src="https://user-images.githubusercontent.com/1676235/75078912-c6d9c700-54bb-11ea-9aa8-8f3e5a13da74.png">
Hey @AmyYLee updating the screenshots, let me know what you think and if I should do more tweaks!

AmyYLee commented:
The micro sub-titles in Settings and sub screens in Settings and in menus should also be in Metropolis.

![Screen Shot 2020-02-25 at 2 34 19 PM](https://user-images.githubusercontent.com/5992746/75280781-1c73e380-57dc-11ea-9d9c-2cf8d9f22b77.png)

AmyYLee commented:
@kglazko Mark-up of where Metropolis needs to be updated in Onboarding. Thanks!

![image](https://user-images.githubusercontent.com/5992746/75576662-60f6bd80-5a2e-11ea-9ed0-d1a9e8ea22e1.png)

AmyYLee commented:
@kglazko 

Here are my edits for Onboarding cards

1. Headers - The letter spacing looks looks wide. Can we reduce the tracking? 
2. Buttons should be Semi-Bold weight (this should apply to all buttons in the app)
3. “Start browsing” button should be in Metropolis Semi-Bold
4. “Automatic” in “Choose your theme” card shouldn’t be Inter
5. Header “Welcome to Firefox Preview” should read “Welcome” - This should be Semi-Bold (looks too bold). There should be a sub header “Already have an account?” - 16pt
6. All headers should be Metropolis Semi-Bold (Looks like Bold)
7. Can we increase the body copy by 1pt? What is it currently? All body copy should be 14pt
8. Can we increase the tracking of the body copy? The text looks tight
9. “Choose your theme” section - The body copy and the “Light theme” and “Dark theme” text look like different sizes. It should be the same size (probably the bigger size)

kglazko commented:
I experimented with making a theme and applying it to preference categories but doing this did not update anything visually. The way recommended online which involves just setting each category to have a layout did work. Note- there is also already a precedence for doing this in the code base as in the preference category for sign in, a layout is being set. So it seems to make sense to do this for the rest @ekager

chrismsimpson commented:
I couldn't help but notice the kerning/spacing in the screenshots above. Metropolis spacing/kerning has been optimised for exactly your use case (on screen user interfaces in the range 12pt - 21pt), so you shouldn't need to set any custom kerning in long form text (as opposed to logo type).

Also, these values will continually be refined, so if you avoid custom kerning and keep your internally packaged version of Metropolis relative up to date downstream you should get spacing & kerning improvements as they are made*.

Here is the current default kerning/spacing values for the latest version of Metropolis Semibold 34pt:

<img width="455" alt="Screen Shot 2020-03-10 at 12 04 55 pm" src="https://user-images.githubusercontent.com/2238758/76269917-92417a00-62c7-11ea-8ceb-53e19303c0fd.png">

*I can already see the kerning pair for lowercase 't' and lowercase 'a' needs work (e.g. 'Start browsing').

kglazko commented:
Thank you so much, @chrismsimpson ! I got rid of the custom spacing and will post an updated screenshot soon.

kglazko commented:
<img width="398" alt="Screen Shot 2020-03-10 at 1 48 45 PM" src="https://user-images.githubusercontent.com/1676235/76358049-0f73eb80-62d6-11ea-8fab-b3edc02754b2.png">
<img width="406" alt="Screen Shot 2020-03-10 at 1 49 01 PM" src="https://user-images.githubusercontent.com/1676235/76358051-11d64580-62d6-11ea-9e94-8b28a43fd66c.png">
@chrismsimpson Is this spacing better?

chrismsimpson commented:
Looks great!

Baron-Severin commented on app/src/main/res/font/metropolis.xml:
```diff
@@ -0,0 +1,85 @@
+            app:fontStyle="normal"
+            app:fontWeight="900"
+            app:font="@font/metropolis_black"
+            />
```
Nit: this isn't formatted according to our style guide, would you mind auto-formatting it?  (put your cursor in the file and press option + cmd + L in AS)

Baron-Severin commented on app/src/main/res/font/metropolis.xml:
```diff
@@ -0,0 +1,85 @@
+            android:fontWeight="900"
+            app:fontStyle="normal"
+            app:fontWeight="900"
+            app:font="@font/metropolis_black"
```
Would you mind adding a comment at the top of this file explaining why we need to duplicate these settings?

Baron-Severin commented on app/src/main/res/font/metropolis.xml:
```diff
@@ -0,0 +1,85 @@
+<?xml version="1.0" encoding="utf-8"?>
+    <font-family xmlns:android="http://schemas.android.com/apk/res/android"
+        xmlns:app="http://schemas.android.com/apk/res-auto">
```
Please add `tools:ignore="UnusedAttribute"` to the `font-family` element.  It'll suppress the lint errors.  The comment mentioned above can sit just below the suppression, explaining it.

Baron-Severin commented on app/src/main/res/layout/account_preference.xml:
```diff
@@ -54,7 +54,9 @@
-            android:textColor="?primaryText" />
+            android:textColor="?primaryText"
+            android:fontFamily="@font/metropolis_bold"
+            android:letterSpacing="0.05" />
```
[This comment](https://github.com/mozilla-mobile/fenix/pull/6986#issuecomment-596852375) seems to suggest that we shouldn't need any custom letter spacing anymore.  Is this one a special case, or was it just accidentally missed?

Baron-Severin commented on app/src/main/res/layout/pbm_shortcut_popup.xml:
```diff
@@ -32,9 +32,12 @@
                 android:textColor="@color/primary_text_dark_theme"
                 android:textSize="16sp"
+                android:fontFamily="@font/metropolis_medium"
+
```
Nit: extra line break

Baron-Severin commented on app/src/main/res/values/styles.xml:
```diff
@@ -214,6 +221,7 @@
         <item name="android:paddingStart">24dp</item>
         <item name="android:paddingEnd">24dp</item>
         <item name="android:elevation">0dp</item>
+        <item name="android:fontFamily">@font/metropolis_semibold</item>
```
Does this one need to be double declared as well?

Baron-Severin commented on app/src/main/res/values/styles.xml:
```diff
@@ -278,15 +288,21 @@
     <style name="ToolbarTitleTextStyle" parent="HeaderTextStyle">
         <item name="android:textSize">20sp</item>
+        <item name="android:fontFamily">@font/metropolis_semibold</item>
+        <item name="fontFamily">@font/metropolis_semibold</item>
```
@ekager apparently `fontFamily` needs a different namespace on different APIs, so we wind up double declaring it all over here.  Do you know if there's any better way to mix it into all of these styles?

ekager replied:
Totally missed this. @kglazko after our chat this week - is this still an issue? Does just using `fontFamily` (using the app namespace) not work on all API levels?

Baron-Severin commented on app/src/main/res/xml/preferences.xml:
```diff
@@ -30,7 +31,9 @@
         android:title="@string/preferences_category_general"
-        app:iconSpaceReserved="false">
+        app:iconSpaceReserved="false"
+        tools:ignore="PrivateResource"
```
Normally, this is a super scary lint warning to suppress.  This is saying that the Android component doesn't consider that field a part of its API, which means they could push an update at any time that makes this code not work.

In _*this*_ case it looks like a false positive though.  It seems to be linting for the filename without regard for the namespace.  If you change the name of `preference_category.xml` those linting errors will go away, and we won't need this suppression.

Btw, it's a good policy to either fix or suppress all lint errors in your PR (and personally if I have to suppress, I try to write a comment explaining why).

Baron-Severin commented on app/src/main/res/xml/preferences.xml:
```diff
@@ -64,7 +67,8 @@
         android:title="@string/preferences_category_privacy_security"
-        app:iconSpaceReserved="false">
+        app:iconSpaceReserved="false"
+        android:layout="@layout/preference_category">
```
It's going to be easy to forget to add this when adding new `PreferenceCategory`s, so I don't love this solution, but the `Preference` API is rough like that sometimes.  If you haven't found any better solutions, we'll have to live with this.

Baron-Severin commented on app/src/main/res/layout/save_tab_group_button.xml:
```diff
@@ -17,8 +17,10 @@
     app:icon="@drawable/ic_tab_collection"
     app:iconGravity="textStart"
     app:iconPadding="8dp"
+    app:fontFamily="@font/metropolis_medium"
```
Some of these views have `fontFamily` declared twice, but others don't.  Are you sure we need this?

kglazko commented on app/src/main/res/layout/account_preference.xml:
```diff
@@ -54,7 +54,9 @@
-            android:textColor="?primaryText" />
+            android:textColor="?primaryText"
+            android:fontFamily="@font/metropolis_bold"
+            android:letterSpacing="0.05" />
```
Nope, good catch!

liuche commented:
@ekager says she'll review this.

ekager commented:
@kglazko could you rebase and update me on the latest status re the android and app namespace question we chatted about last week? And then I will review so we can finally get this landed!

kglazko commented:
Hi Emily! Rebased :) So, I changed it to app in any styling that used app, and utilized android: in those where the rest of the styling was android: (avoiding making it the only app: in the view, maybe that's right, but that can be up to you to decide). In Styles, making it only use app namespace.

ekager commented:
Could we change to always using the `app` namespace when available (ie always :) )? Let's just be consistent and this _should_ always use the correct support version to be more compatible.

kglazko commented:
Absolutely, changed them!
